x-pack/test/case_api_integration/basic/tests/cases/find_cases.ts

```diff
@@ -6,246 +6,454 @@
+          .get(`${CASES_URL}/_find`)
+          .query({
+            sortOrder: 'asc',
+            page: 1,
+            perPage: 11,
+          })
+          .set('kbn-xsrf', 'true')
+          .expect(200);
+
+        expect(body.total).to.eql(10);
+        expect(body.page).to.eql(1);
+        expect(body.per_page).to.eql(11);
+        expect(body.cases.length).to.eql(10);
+        expect(body.count_open_cases).to.eql(10);
+        expect(body.count_closed_cases).to.eql(0);
+        expect(body.count_in_progress_cases).to.eql(0);
+      });
 
-      await supertest
-        .patch(CASES_URL)
-        .set('kbn-xsrf', 'true')
-        .send({
-          cases: [
-            {
-              id: postedCase.id,
-              version: postedCase.version,
-              status: 'closed',
-            },
-          ],
-        })
-        .expect(200);
+      it('returns the correct total when perPage is equal to the total', async () => {
+        const { body }: { body: CasesFindResponse } = await supertest
+          .get(`${CASES_URL}/_find`)
+          .query({
+            sortOrder: 'asc',
+            page: 1,
+            perPage: 10,
+          })
+          .set('kbn-xsrf', 'true')
+          .expect(200);
+
+        expect(body.total).to.eql(10);
+        expect(body.page).to.eql(1);
+        expect(body.per_page).to.eql(10);
+        expect(body.cases.length).to.eql(10);
+        expect(body.count_open_cases).to.eql(10);
+        expect(body.count_closed_cases).to.eql(0);
+        expect(body.count_in_progress_cases).to.eql(0);
+      });
 
-      await supertest
-        .patch(CASES_URL)
-        .set('kbn-xsrf', 'true')
-        .send({
-          cases: [
-            {
-              id: inProgreeCase.id,
-              version: inProgreeCase.version,
-              status: 'in-progress',
-            },
-          ],
-        })
-        .expect(200);
-
-      const { body } = await supertest
-        .get(`${CASES_URL}/_find?sortOrder=asc`)
-        .set('kbn-xsrf', 'true')
-        .send()
-        .expect(200);
-
-      expect(body.count_open_cases).to.eql(1);
-      expect(body.count_closed_cases).to.eql(1);
-      expect(body.count_in_progress_cases).to.eql(1);
-    });
+      it('returns the second page of results', async () => {
+        const perPage = 5;
+        const { body }: { body: CasesFindResponse } = await supertest
+          .get(`${CASES_URL}/_find`)
+          .query({
+            sortOrder: 'asc',
+            page: 2,
+            perPage,
+          })
+          .set('kbn-xsrf', 'true')
+          .expect(200);
+
+        expect(body.total).to.eql(10);
+        expect(body.page).to.eql(2);
+        expect(body.per_page).to.eql(5);
+        expect(body.cases.length).to.eql(5);
+        expect(body.count_open_cases).to.eql(10);
+        expect(body.count_closed_cases).to.eql(0);
+        expect(body.count_in_progress_cases).to.eql(0);
+
+        const allCases = await getAllCasesSortedByCreatedAtAsc();
```

Comment:
I read your comment on top of this function but still I don not get why we need this to validate the cases?

Reply:
Sorry for the confusion, I'll see if I can describe issue here. Essentially, for the tests that walk the pages I wanted some way to verify that the cases returned by the `_find` API were in the correct order (and actually on that page). To do that we need a reliable way to retrieve all the cases and then we can compare the _chunk_ that the `_find` API returns for a specific page.

For example if the test requests page 2 with a per_page of 2 when there are 10 total cases (say case, 0,1,2,3,4,5 etc created in that order), then the find API retrieve cases 2 and 3 (0 and 1 would be on the _first_ page). So we need a source of truth ideally separate from the find API to gather all the cases and then compare the find results against that ground truth.

So then the question is how can we ensure that specific cases will be on a specific page for pagination. We could make 10 create API requests one at a time that each create a case. This would guarantee that the `created_at` field of the 10 cases would be in ascending order and we could get away with just comparing the find response's `created_at` to the response from the create API in the test setup. Unfortunately, creating cases one at a time was pretty slow.

Another option would be not have the sorting be on the `created_at` field. For example if we could set each case `title` to a deterministic value as we create the tests then we'd know what the order would be when the find API is paginating through them. Unfortunately, the case backend implementation only allows us to sort on a few fields and `title` isn't allowed.

So the only _fast_ solution I could think of was to create the cases using a `Promise.all` and use an _outside_ method (using Elasticsearch directly to retrieve the cases in sorted order) to retrieve all the cases. `Promise.all` doesn't guarantee that for a specific `i` in the loop that the case's `created_at` date would be in order because they're being done asynchronously. The loop goes from 0-9 but case 9 could be created before 0 for example.

Reply:
I see! Now all make sense. Thanks for the detailed explanation!